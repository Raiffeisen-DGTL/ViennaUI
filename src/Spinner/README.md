# Spinner

Визуализация состояния загрузки данных. Используется на страницах или в компонентах (например в компонентах `<Input />` или `<Select />`), когда нужно визуализировать загрузку данных с сервера.

## Импорт

```
import { Spinner } from 'vienna-ui';
```

## Свойства / Props

Свойства наследуются от [HTMLAttributes<HTMLDivElement>](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L1746)

| Prop     | Type                                                    | Default | Description |
| -------- | ------------------------------------------------------- | ------- | ----------- |
| size     | ResponsiveProp<"xs" \| "s" \| "m" \| "l" \| "xl" \| "xxl", B extends void ? Breakpoints : B> \| undefined |    |
| position | "relative" \| "absolute" \| undefined                   |    |
| color | 'primary' \| 'accent' \| 'london120' \| 'white' \| undefined                 |    |


# Spinner

Визуализация состояния загрузки данных.

Используется на страницах или в компонентах (например в компонентах `<Input />` или `<Select />`), когда нужно визуализировать загрузку данных данных с сервера.



```
    <Spinner />
```

## Внешний вид

```
    <Spinner />
```

## Размеры

Компонент имеет стандартные размеры `xs`, `s`, `m`, `l` и `xl`.

```
    <Spinner size='xs' />
    <Spinner size='s' />
    <Spinner size='m' />
    <Spinner size='l' />
    <Spinner size='xl' />
```

## Color

#### Primary

По умолчанию

```
    <Spinner size='xs' color='primary' />
    <Spinner size='s' color='primary' />
    <Spinner size='m' color='primary' />
    <Spinner size='l' color='primary' />
    <Spinner size='xl' color='primary' />
```

#### Accent

```
    <Spinner size='xs' color='accent' />
    <Spinner size='s' color='accent' />
    <Spinner size='m' color='accent' />
    <Spinner size='l' color='accent' />
    <Spinner size='xl' color='accent' />
```

#### London120

```
    <Spinner size='xs' color='london120' />
    <Spinner size='s' color='london120' />
    <Spinner size='m' color='london120' />
    <Spinner size='l' color='london120' />
    <Spinner size='xl' color='london120' />
```

#### White

```
    <Spinner size='xs' color='white' />
    <Spinner size='s' color='white' />
    <Spinner size='m' color='white' />
    <Spinner size='l' color='white' />
    <Spinner size='xl' color='white' />
```

## Расположение

Можно добавить свойство `position` в состояние `absolute` для растягивания на всю ширину или высоту родительского элемента.

```
    <Spinner position='absolute' />
```

#### Адаптив

Для компонента Spinner, адаптив применяется к свойству `size`, что позволяет адаптивно менять размер компонента в зависимости от текущей ширины экрана. Для этого задайте свойству `size` объект вида `{ <breakpoint name>: <string value> }`

Основные значения breakpoints:

```
defaultBreakpoints = {
    s: 768,
    m: 1024,
    l: 1920,
    xl: 2560,
};

systemBreakpoints: Breakpoints = {
    /* xs */
    xs: `(max-width: ${defaultBreakpoints.s - 1}px)`,

    /* s */
    s: `(min-width: ${defaultBreakpoints.s}px)`,
    belowS: `(max-width: ${defaultBreakpoints.s - 1}px)`,

    /* m */
    m: `(min-width: ${defaultBreakpoints.m}px)`,
    belowM: `(max-width: ${defaultBreakpoints.m - 1}px)`,

    /* l */
    l: `(min-width: ${defaultBreakpoints.l}px)`,
    belowL: `(max-width: ${defaultBreakpoints.l - 1}px)`,

    /* xl */
    xl: `(min-width: ${defaultBreakpoints.xl}px)`,
};
```

```
    <Spinner size={{ base: 's', s: 'l', m: 'xl' }} />
```

## Установка data-testid

Атрибут `data-testid` можно передать для `menu`. Передается пропс ` testId?: {container};`.

```
    <Spinner testId={{container: 'test'}} size={{ base: 's', s: 'l', m: 'xl' }} />
```