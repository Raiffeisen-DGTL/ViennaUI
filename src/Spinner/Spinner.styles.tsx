import styled, { css, keyframes } from 'styled-components';
import { spinner } from 'vienna.ui-theme';
import { getPresets } from '../Utils/styling';
import { ResponsiveProp, Breakpoints } from '../Utils/responsiveness';

const presets = getPresets(
    spinner,
    'spinner'
)({
    size: ['$size', 'm'],
    color: '$color',
});

const rotateIE360 = keyframes`
  from {
    transform: rotate(0);
  }

  to {
    transform: rotate(360deg);
  }
`;

const rotate360 = keyframes`
    0% {
        stroke-dasharray: 85 195;
        stroke-dashoffset: 0;
      }
      10% {
        stroke-dasharray: 100 180;
        stroke-dashoffset: -60;
      }
      30% {
        stroke-dasharray: 260 20;
        stroke-dashoffset: -290;
      }
      50% {
        stroke-dasharray: 85 195;
        stroke-dashoffset: -580;
      }
      60% {
        stroke-dasharray: 85 195;
        stroke-dashoffset: -615;
      }
      63% {
        stroke-dasharray: 85 195;
        stroke-dashoffset: -620;
      }
      70% {
        stroke-dasharray: 85 195;
        stroke-dashoffset: -635;
      }
      80% {
        stroke-dasharray: 85 195;
        stroke-dashoffset: -560;
      }
      84% {
        stroke-dasharray: 85 195;
        stroke-dashoffset: -545;
      }
      98% {
        stroke-dasharray: 85 195;
        stroke-dashoffset: -530;
      }
      100% {
        stroke-dasharray: 85 195;
        stroke-dashoffset: -530;
      }
`;

interface Size<B = Breakpoints> {
    $size: ResponsiveProp<'xs' | 's' | 'm' | 'l' | 'xl' | 'xxl', B>;
}

export interface PropsBox<B = Breakpoints> extends Size<B> {
    $position: 'relative' | 'absolute';
}
export const Box = styled.div<PropsBox>`
    background-size: cover;
    background-repeat: no-repeat;
    vertical-align: top;
    position: relative;
    display: inline-flex;
    ${presets.size};

    ${({ $position }) =>
        $position === 'absolute' &&
        css`
            position: absolute;
            top: 0;
            left: 0;
            bottom: 0;
            right: 0;
            z-index: 5;
            margin: auto;
        `};
`;

interface PropsSvg<B = Breakpoints> extends Size<B> {
    $isIE: boolean;
}
export const Svg = styled.svg<PropsSvg>`
    ${presets.size}
    ${({ $isIE }) =>
        $isIE &&
        css`
            animation: ${rotateIE360} ease 1s infinite;
        `}
    display: block;
`;

export interface CircleProps {
    $color: 'primary' | 'accent' | 'london120' | 'white' | 'seattle60';
    $isIE: boolean;
}
export const Circle = styled.circle<CircleProps>`
    stroke-width: 12px;
    stroke-miterlimit: 10;
    fill: none;
    stroke-dasharray: ${({ $isIE }) => ($isIE ? `195 85` : `85 195`)};
    stroke-dashoffset: 0;
    stroke-linecap: round;
    transform-origin: 50%;
    ${({ $isIE }) =>
        !$isIE &&
        css`
            animation: ${rotate360} linear 2s infinite;
        `}

    ${presets.color}
`;
