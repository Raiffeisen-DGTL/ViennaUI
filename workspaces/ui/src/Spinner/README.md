# Spinner

Визуализация состояния загрузки данных.
Используется на страницах или в компонентах (например в компонентах `<Input />` или `<Select />`), когда нужно визуализировать загрузку данных с сервера.


## Импорт

```
import { Spinner } from 'vienna-ui';
```

## Свойства / Props

Свойства наследуются от [HTMLAttributes<HTMLDivElement>](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L1746)

Prop | Type | Default | Description
--- | --- | --- | ---
size | "xs" \| "s" \| "m" \| "l" \| "xl" \| "xxl" \| undefined | false |
position | "relative" \| "absolute" \| undefined | false |

## Использование

```<Spinner />```

## Размеры
##### Свойство `size`

Компонент имеет стандартные размеры `size`: `xs`, `s`, `m`, `l` и `xl`.

## Расположение
##### Свойство `position`

Можно добавить свойство `position` в состояние `absolute` для растягивания на всю ширину или высоту родительского элемента.

```<Spinner position='absolute' />```
