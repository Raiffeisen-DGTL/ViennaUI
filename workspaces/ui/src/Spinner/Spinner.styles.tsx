import styled, { css, keyframes } from 'styled-components';
import { getPresets } from 'vienna.ui-primitives';

const { size, color } = getPresets('spinner', {
    size: 'size',
    color: 'color',
});

const rotateIE360 = keyframes`
  from {
    transform: rotate(0);
  }

  to {
    transform: rotate(360deg);
  }
`;

const rotate360 = keyframes`
    0% {
        stroke-dasharray: 85 195;
        stroke-dashoffset: 0;
      }
      10% {
        stroke-dasharray: 100 180;
        stroke-dashoffset: -60;
      }
      30% {
        stroke-dasharray: 260 20;
        stroke-dashoffset: -290;
      }
      50% {
        stroke-dasharray: 85 195;
        stroke-dashoffset: -580;
      }
      60% {
        stroke-dasharray: 85 195;
        stroke-dashoffset: -615;
      }
      63% {
        stroke-dasharray: 85 195;
        stroke-dashoffset: -620;
      }
      70% {
        stroke-dasharray: 85 195;
        stroke-dashoffset: -635;
      }
      80% {
        stroke-dasharray: 85 195;
        stroke-dashoffset: -560;
      }
      84% {
        stroke-dasharray: 85 195;
        stroke-dashoffset: -545;
      }
      98% {
        stroke-dasharray: 85 195;
        stroke-dashoffset: -530;
      }
      100% {
        stroke-dasharray: 85 195;
        stroke-dashoffset: -530;
      }
`;

export const Box = styled.div<{
    [key: string]: any;
    size: 'xs' | 's' | 'm' | 'l' | 'xl' | 'xxl';
    position: 'relative' | 'absolute';
}>`
    background-size: cover;
    background-repeat: no-repeat;
    vertical-align: top;
    position: relative;
    display: inline-flex;
    ${size};

    ${(props): any =>
        props.position === 'absolute' &&
        css`
            position: absolute;
            top: 0;
            left: 0;
            bottom: 0;
            right: 0;
            z-index: 5;
            margin: auto;
        `};
`;

export const Svg = styled.svg<{ size: 'xs' | 's' | 'm' | 'l' | 'xl' | 'xxl'; isIE: boolean }>`
    ${size}
    ${({ isIE }) =>
        isIE &&
        css`
            animation: ${rotateIE360} ease 1s infinite;
        `}
    display: block;
`;

export const Circle = styled.circle<{ color: 'primary' | 'accent' | 'london120' | 'white'; isIE: boolean }>`
    stroke-width: 12px;
    stroke-miterlimit: 10;
    fill: none;
    stroke-dasharray: ${({ isIE }) => (isIE ? `195 85` : `85 195`)};
    stroke-dashoffset: 0;
    stroke-linecap: round;
    transform-origin: 50%;
    ${({ isIE }) =>
        !isIE &&
        css`
            animation: ${rotate360} linear 2s infinite;
        `}

    ${color}
`;
