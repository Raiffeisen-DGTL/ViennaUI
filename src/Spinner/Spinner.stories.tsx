import React from 'react';
import { Story, Meta } from 'storybook';
import { Spinner, SpinnerProps } from './Spinner';

export default {
    title: 'Development/Spinner',
    component: Spinner,
} as Meta;

export const Overview: Story<SpinnerProps> = (args) => {
    return <Spinner {...args} />;
};
