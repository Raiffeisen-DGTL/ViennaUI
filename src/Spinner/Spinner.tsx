import React, { useMemo } from 'react';
import { Box, Svg, Circle, CircleProps, PropsBox } from './Spinner.styles';
import { Breakpoints } from '../Utils/responsiveness';
import { BoxStyled } from '../Utils/styled';

export interface SpinnerProps<B = Breakpoints> extends BoxStyled<typeof Box, PropsBox> {
    size?: PropsBox<B>['$size'];
    position?: PropsBox<B>['$position'];
    color?: CircleProps['$color'];
}

export function Spinner<B = void>(props: SpinnerProps<B extends void ? Breakpoints : B>) {
    const { size = 'm', position = 'relative', color = 'primary', ...attrs } = props;

    const isIE = useMemo(() => {
        const isBrowser = typeof window !== 'undefined';
        if (isBrowser) {
            const ua = window.navigator.userAgent;
            return ua.indexOf('MSIE ') > 0 || ua.indexOf('Trident/') > 0;
        }

        return false;
    }, []);

    return (
        <Box {...(attrs as {})} $size={size} $position={position}>
            <Svg $isIE={isIE} viewBox='0 0 100 100' xmlns='http://www.w3.org/2000/svg' $size={size}>
                <Circle $isIE={isIE} r='44' cx='50' cy='50' $color={color} />
            </Svg>
        </Box>
    );
}

Spinner.displayName = 'Spinner';
