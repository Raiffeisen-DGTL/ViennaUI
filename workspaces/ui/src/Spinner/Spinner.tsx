import React, { HTMLAttributes } from 'react';
import { Box, Svg } from './Spinner.styles';

export interface SpinnerProps extends Omit<HTMLAttributes<HTMLDivElement>, 'size'> {
    size?: 'xs' | 's' | 'm' | 'l' | 'xl' | 'xxl';
    position?: 'relative' | 'absolute';
}

export const Spinner: React.FC<SpinnerProps> = (props): JSX.Element => {
    const { size = 'm', position = 'relative', ...attrs } = props;

    return (
        <Box {...attrs} size={size} position={position}>
            <Svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16' size={size}>
                <g fill='none' fillRule='evenodd' strokeWidth='2'>
                    <path stroke='#2b2d33' strokeOpacity='.87' d='M3.05 12.95a7 7 0 0 1 9.9-9.9' />
                    <path stroke='#fee600' d='M3.05 12.95a7 7 0 0 0 9.9-9.9' />
                </g>
            </Svg>
        </Box>
    );
};

Spinner.displayName = 'Spinner';
Spinner.defaultProps = {
    size: 'm',
    position: 'relative',
};

export default Spinner;
