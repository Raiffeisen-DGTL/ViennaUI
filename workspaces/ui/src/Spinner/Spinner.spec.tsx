import React from 'react';
import Spinner, { SpinnerProps } from './Spinner';

test('Spinner', () => {
    const snap = snapshot.render(<Spinner />);
    expect(snap).toMatchSnapshot();
});

test('Spinner w/ size', () => {
    const sizes: SpinnerProps['size'][] = ['xs', 's', 'm', 'l', 'xl', 'xxl'];

    sizes.forEach((size) => {
        const snap = snapshot.render(<Spinner size={size} />);
        expect(snap).toMatchSnapshot(size);
    });
});

test('Spinner w/ position', () => {
    const positions: SpinnerProps['position'][] = ['relative', 'absolute'];

    positions.forEach((position) => {
        const snap = snapshot.render(<Spinner position={position} />);
        expect(snap).toMatchSnapshot(position);
    });
});

test('Spinner w/ any props', () => {
    const props = { className: 'className', id: 'Spinner', title: 'title', onClick: () => 'Spinner' };
    const component = <Spinner {...props} />;

    const snap = snapshot.render(component);
    expect(snap).toMatchSnapshot();

    const wrapper = snapshot.shallow(component);
    for (const [key, value] of Object.entries(props)) {
        expect(wrapper.props()[key]).toEqual(value);
    }
});
