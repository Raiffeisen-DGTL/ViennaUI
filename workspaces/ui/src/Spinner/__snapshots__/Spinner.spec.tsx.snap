// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner 1`] = `
<div
  class="Spinnerstyles__Box-moemrn-0 jPiLVW"
>
  <svg
    class="Spinnerstyles__Svg-moemrn-1 cPXsgS"
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      class="Spinnerstyles__Circle-moemrn-2 cElquL"
      color="primary"
      cx="50"
      cy="50"
      r="44"
    />
  </svg>
</div>
`;

exports[`Spinner w/ any props 1`] = `
<div
  class="Spinnerstyles__Box-moemrn-0 jPiLVW className"
  id="Spinner"
  title="title"
>
  <svg
    class="Spinnerstyles__Svg-moemrn-1 cPXsgS"
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      class="Spinnerstyles__Circle-moemrn-2 cElquL"
      color="primary"
      cx="50"
      cy="50"
      r="44"
    />
  </svg>
</div>
`;

exports[`Spinner w/ position: absolute 1`] = `
<div
  class="Spinnerstyles__Box-moemrn-0 eAchHT"
>
  <svg
    class="Spinnerstyles__Svg-moemrn-1 cPXsgS"
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      class="Spinnerstyles__Circle-moemrn-2 cElquL"
      color="primary"
      cx="50"
      cy="50"
      r="44"
    />
  </svg>
</div>
`;

exports[`Spinner w/ position: relative 1`] = `
<div
  class="Spinnerstyles__Box-moemrn-0 jPiLVW"
>
  <svg
    class="Spinnerstyles__Svg-moemrn-1 cPXsgS"
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      class="Spinnerstyles__Circle-moemrn-2 cElquL"
      color="primary"
      cx="50"
      cy="50"
      r="44"
    />
  </svg>
</div>
`;

exports[`Spinner w/ size: l 1`] = `
<div
  class="Spinnerstyles__Box-moemrn-0 jDmrri"
>
  <svg
    class="Spinnerstyles__Svg-moemrn-1 lhJvVO"
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      class="Spinnerstyles__Circle-moemrn-2 cElquL"
      color="primary"
      cx="50"
      cy="50"
      r="44"
    />
  </svg>
</div>
`;

exports[`Spinner w/ size: m 1`] = `
<div
  class="Spinnerstyles__Box-moemrn-0 jPiLVW"
>
  <svg
    class="Spinnerstyles__Svg-moemrn-1 cPXsgS"
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      class="Spinnerstyles__Circle-moemrn-2 cElquL"
      color="primary"
      cx="50"
      cy="50"
      r="44"
    />
  </svg>
</div>
`;

exports[`Spinner w/ size: s 1`] = `
<div
  class="Spinnerstyles__Box-moemrn-0 eAFeS"
>
  <svg
    class="Spinnerstyles__Svg-moemrn-1 lbgYES"
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      class="Spinnerstyles__Circle-moemrn-2 cElquL"
      color="primary"
      cx="50"
      cy="50"
      r="44"
    />
  </svg>
</div>
`;

exports[`Spinner w/ size: xl 1`] = `
<div
  class="Spinnerstyles__Box-moemrn-0 gSyAmC"
>
  <svg
    class="Spinnerstyles__Svg-moemrn-1 chftsm"
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      class="Spinnerstyles__Circle-moemrn-2 cElquL"
      color="primary"
      cx="50"
      cy="50"
      r="44"
    />
  </svg>
</div>
`;

exports[`Spinner w/ size: xs 1`] = `
<div
  class="Spinnerstyles__Box-moemrn-0 qwZJG"
>
  <svg
    class="Spinnerstyles__Svg-moemrn-1 cJuUPW"
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      class="Spinnerstyles__Circle-moemrn-2 cElquL"
      color="primary"
      cx="50"
      cy="50"
      r="44"
    />
  </svg>
</div>
`;

exports[`Spinner w/ size: xxl 1`] = `
<div
  class="Spinnerstyles__Box-moemrn-0 gSyAmC"
>
  <svg
    class="Spinnerstyles__Svg-moemrn-1 chftsm"
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      class="Spinnerstyles__Circle-moemrn-2 cElquL"
      color="primary"
      cx="50"
      cy="50"
      r="44"
    />
  </svg>
</div>
`;
